graphql/schemabuilder/batch.go

```diff
@@ -0,0 +1,300 @@
+		return nil, fmt.Errorf("batch Field funcs require at least one input field")
+	}
+
+	in = funcCtx.consumeContext(in)
+	in, err = funcCtx.consumeRequiredSourceBatch(in)
+	if err != nil {
+		return nil, err
+	}
+	argParser, args, in, err := funcCtx.consumeArgs(sb, in)
+	if err != nil {
+		return nil, err
+	}
+	in = funcCtx.consumeSelectionSet(in)
+
+	// We have succeeded if no arguments remain.
+	if len(in) != 0 {
+		return nil, fmt.Errorf("%s arguments should be [context,]map[int][*]%s[, args][, selectionSet]", funcCtx.funcType, typ)
+	}
+
+	out := funcCtx.getFuncOutputTypes()
+	retType, out, err := funcCtx.consumeReturnValue(m, sb, out)
+	if err != nil {
+		return nil, err
+	}
+	out = funcCtx.consumeReturnError(out)
+	if len(out) > 0 {
+		return nil, fmt.Errorf("%s return should be [map[int]<Type>][,error]", funcCtx.funcType)
+	}
+
+	return &graphql.Field{
+		Resolve: func(ctx context.Context, source, funcRawArgs interface{}, selectionSet *graphql.SelectionSet) (interface{}, error) {
+			// TODO use real batching.
+			sources := []interface{}{source}
+			// Set up function arguments.
+			funcInputArgs := funcCtx.prepareResolveArgs(sources, funcRawArgs, ctx, selectionSet)
+
+			// Call the function.
+			funcOutputArgs := callableFunc.Call(funcInputArgs)
+
+			results, err := funcCtx.extractResultsAndErr(len(sources), funcOutputArgs, retType)
+			if err != nil {
+				return nil, err
+			}
+			return results[0], nil
+		},
+		Args:           args,
+		Type:           retType,
+		ParseArguments: argParser.Parse,
+		Expensive:      funcCtx.hasContext,
+	}, nil
+}
+
+// funcContext is used to parse the function signature in buildFunction.
+type batchFuncContext struct {
+	hasContext      bool
+	hasSource       bool
+	hasArgs         bool
+	hasSelectionSet bool
+	hasRet          bool
+	hasError        bool
+
+	funcType     reflect.Type
+	batchMapType reflect.Type
+	isPtrFunc    bool
+	parentTyp    reflect.Type
+}
+
+// getFuncVal returns a reflect.Value of an executable function.
+func (funcCtx *batchFuncContext) getFuncVal(m *method) (reflect.Value, error) {
+	fun := reflect.ValueOf(m.Fn)
+	if fun.Kind() != reflect.Func {
+		return fun, fmt.Errorf("fun must be func, not %s", fun)
+	}
+	funcCtx.funcType = fun.Type()
+	return fun, nil
+}
+
+// getFuncInputTypes returns the input arguments for the function we're
+// representing.
+func (funcCtx *batchFuncContext) getFuncInputTypes() []reflect.Type {
+	in := make([]reflect.Type, 0, funcCtx.funcType.NumIn())
+	for i := 0; i < funcCtx.funcType.NumIn(); i++ {
+		in = append(in, funcCtx.funcType.In(i))
+	}
+	return in
+}
+
+// consumeContext reads in the input parameters for the provided
+// function and determines whether the function has a Context input parameter.
+// It returns the input types without the context parameter if it was there.
+func (funcCtx *batchFuncContext) consumeContext(in []reflect.Type) []reflect.Type {
+	if len(in) > 0 && in[0] == contextType {
+		funcCtx.hasContext = true
```

Comment:
I'd make this function return (hasContext, in[1:]) so that the caller is forced to store hasContext. i like functions that don't modify state, if possible.